test/e2e/route_metrics_test.go

```diff
@@ -0,0 +1,583 @@
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	ic.Spec.RouteSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"type": "rs-foo",
+		},
+	}
+
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+
+	// Delete the IC if any error occurs.
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	// Store the start time to get the time taken to update metrics after the creation of the IC.
+	startTime := time.Now()
+
+	// Wait for metrics to be added and set to 0.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	// Get the time taken for the metrics to be updated.
+	t.Logf("time taken for the metrics to be updated after the creation of the IC: %fs", time.Since(startTime).Seconds())
+
+	// Create a Route to be immediately admitted by this Ingress Controller.
+	// Use openshift-console namespace to get a namespace outside the ingress-operator's cache.
+	routeFooLabelName := types.NamespacedName{Namespace: "openshift-console", Name: "route-rs-foo-label"}
+	routeFooLabel := newRouteWithLabel(routeFooLabelName, "rs-foo")
+	if err := kclient.Create(context.TODO(), routeFooLabel); err != nil {
+		t.Fatalf("failed to create route: %v", err)
+	}
+
+	// Delete the Route if any error occurs.
+	defer func() {
+		if err := kclient.Delete(context.TODO(), routeFooLabel); err != nil {
+			if apierrors.IsNotFound(err) {
+				return
+			} else {
+				t.Fatalf("failed to delete route %s: %v", routeFooLabelName, err)
+			}
+		}
+		t.Logf("deleted route: %s", routeFooLabelName)
+	}()
+
+	// Wait for metrics to be updated to 1 as the Route will get admitted by the IC.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 1); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	// Update the RouteSelector of the Ingress Controller so that the Route gets un-admitted.
+	if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+		t.Fatalf("failed to get ingress resource: %v", err)
+	}
+	ic.Spec.RouteSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"type": "rs-bar",
+		},
+	}
+	if err := kclient.Update(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	// Wait for metrics to be updated to zero as the Route will get un-admitted by the IC.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	// Update the RouteSelector of the Ingress Controller so that the Route gets admitted again.
+	if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+		t.Fatalf("failed to get ingress resource: %v", err)
+	}
+	ic.Spec.RouteSelector = &metav1.LabelSelector{
```

Comment:
Do you think we should also test changing the route's label to un-admit itself from a IC? That should actually work right? Because we are leaving the logic for filter via route labels and you reconcile on a route update.

However, when it comes to updating the namespace label to un-admit a route, that still won't work, but should work after we fix https://issues.redhat.com/browse/OCPBUGS-1689, because it will modify the route's status and cause a reconciliation.